homeassistant/components/modbus/__init__.py

```diff
@@ -87,13 +104,52 @@
     }
 )
 
+COVERS_SCHEMA = vol.All(
+    cv.has_at_least_one_key(CALL_TYPE_COIL, CONF_REGISTER),
+    vol.Schema(
+        {
+            vol.Required(CONF_HUB): cv.string,
+            vol.Required(CONF_NAME): cv.string,
+            vol.Optional(
+                CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+            ): cv.positive_int,
+            vol.Optional(CONF_DEVICE_CLASS): COVER_DEVICE_CLASSES_SCHEMA,
+            vol.Optional(CONF_SLAVE, default=DEFAULT_SLAVE): cv.positive_int,
+            vol.Optional(CONF_STATE_CLOSED, default=0): cv.positive_int,
+            vol.Optional(CONF_STATE_CLOSING, default=3): cv.positive_int,
+            vol.Optional(CONF_STATE_OPEN, default=1): cv.positive_int,
+            vol.Optional(CONF_STATE_OPENING, default=2): cv.positive_int,
+            vol.Optional(CONF_STATUS_REGISTER): cv.positive_int,
+            vol.Optional(
+                CONF_STATUS_REGISTER_TYPE, default=CALL_TYPE_REGISTER_HOLDING,
+            ): vol.In([CALL_TYPE_REGISTER_HOLDING, CALL_TYPE_REGISTER_INPUT]),
+            vol.Exclusive(CALL_TYPE_COIL, CONF_INPUT_TYPE): cv.positive_int,
+            vol.Exclusive(CONF_REGISTER, CONF_INPUT_TYPE): cv.positive_int,
+        }
+    ),
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_HUBS): vol.All(
+                    cv.ensure_list, [vol.Any(SERIAL_SCHEMA, ETHERNET_SCHEMA)]
+                ),
+                vol.Optional(CONF_COVERS): vol.All(cv.ensure_list, [COVERS_SCHEMA]),
```

Comment:
Maybe it's better to put the cover schema inside the base hub schema. Then we won't make a breaking change and the covers will be coupled naturally to a hub.

Reply:
I thought about this variant, too, and I think it's better this way. Covers list is defined only once, and each entry points to proper Modbus hub. To me it looks concise. Also, I have changes planned for `light` and `fan` entities (new integrations). Having all lights, covers and fans defined in one place is in my opinion much more readable.

Consider this:
```yaml
modbus:
  - name: plc_01
    type: tcp
    host: 192.168.1.31
    port: 502
    covers:
        - name: plc_01_door
          hub: plc_01
          ....
    lights:
        - name: plc_01_light
           hub: plc_01
          ...
    fans:
        - name: plc_01_fan
           hub: plc_01
          ...
    name: plc_02
    type: tcp
    host: 192.168.1.32
    port: 502
    covers:
        - name: plc_02_door
          hub: plc_02
          ....
    lights:
        - name: plc_02_light
           hub: plc_02
          ...
    fans:
        - name: plc_02_fan
           hub: plc_02
          ...
````

Versus this:
```yaml
modbus:
    hubs:
      - name: plc_01
        type: tcp
        host: 192.168.1.31
        port: 502
      - name: plc_02
        type: tcp
        host: 192.168.1.32
        port: 502

    covers:
        - name: plc_01_door
          hub: plc_01
          ....
        - name: plc_02_door
          hub: plc_02
          ....

    lights:
        - name: plc_01_light
           hub: plc_01
          ...
        - name: plc_02_light
           hub: plc_02
          ...

    fans:
        - name: plc_01_fan
           hub: plc_01
          ...
        - name: plc_02_fan
           hub: plc_02
          ...
```

Second config is much better readable, especially if there area more than two hubs.

Reply:
We don't need to define hub under the covers etc in the first variant. I think that's better and we're avoiding a breaking change. The latter is very important.

Reply:
I can always write a compatibility code, if this is the only problem.

- If people don't use the new `cover` component, their config will still work. I'll just print a warning.
- If they decide to use the new `cover` component, they need to rework their config anyway, and instructions already show what needs to be changed

Reply:
We'll break all existing user configs since we have changed the config schema. We shouldn't do that if we don't need to. Please use the suggestion. That's also in line with other integrations.